src/subsystems.jl

```diff
@@ -0,0 +1,283 @@
+    IS.add_subsystem!(sys.data, subsystem_name)
+end
+
+"""
+Return the number of subsystems stored in the system.
+"""
+get_num_subsystems(sys::System) = IS.get_num_subsystems(sys.data)
+
+"""
+Return an iterator of all subsystem names in the system.
+"""
+get_subsystems(sys::System) = IS.get_subsystems(sys.data)
+
+"""
+Remove a subsystem from the system.
+
+Throws ArgumentError if the subsystem name is not stored.
+"""
+remove_subsystem!(sys::System, subsystem_name::AbstractString) =
+    IS.remove_subsystem!(sys.data, subsystem_name)
+
+"""
+Return true if the system has one or more subsystems.
+"""
+function has_subsystems(sys::System)
+    for _ in get_subsystems(sys)
+        return true
+    end
+
+    return false
+end
+
+"""
+Add a component to a subsystem.
+"""
+function add_component_to_subsystem!(
+    sys::System,
+    subsystem_name::AbstractString,
+    component::Component,
+)
+    IS.add_component_to_subsystem!(sys.data, subsystem_name, component)
+    handle_component_addition_to_subsystem!(sys, subsystem_name, component)
+    return
+end
+
+"""
+Peforms component-type-specific postprocessing when a component is added to a subsystem.
+"""
+handle_component_addition_to_subsystem!(
+    ::System,
+    subsystem_name::AbstractString,
+    ::Component,
+) = nothing
+
+"""
+Peforms component-type-specific postprocessing when a component is removed from a subsystem.
+"""
+handle_component_removal_from_subsystem!(
+    ::System,
+    subsystem_name::AbstractString,
+    ::Component,
+) = nothing
+
+function handle_component_addition_to_subsystem!(
+    sys::System,
+    subsystem_name::AbstractString,
+    component::StaticInjectionSubsystem,
+)
+    for subcomponent in get_subcomponents(component)
+        if !is_assigned_to_subsystem(sys, subcomponent, subsystem_name)
+            add_component_to_subsystem!(sys, subsystem_name, subcomponent)
+        end
+    end
+end
+
+function handle_component_removal_from_subsystem!(
+    sys::System,
+    subsystem_name::AbstractString,
+    component::StaticInjectionSubsystem,
+)
+    for subcomponent in get_subcomponents(component)
+        if is_assigned_to_subsystem(sys, subcomponent, subsystem_name)
+            remove_component_from_subsystem!(sys, subsystem_name, subcomponent)
+        end
+    end
+end
+
+function handle_component_addition_to_subsystem!(
+    sys::System,
+    subsystem_name::AbstractString,
+    component::RegulationDevice,
+)
+    if !is_assigned_to_subsystem(sys, component.device, subsystem_name)
+        add_component_to_subsystem!(sys, subsystem_name, component.device)
+    end
+end
+
+function handle_component_removal_from_subsystem!(
+    sys::System,
+    subsystem_name::AbstractString,
+    component::RegulationDevice,
+)
+    if is_assigned_to_subsystem(sys, component.device, subsystem_name)
+        remove_component_from_subsystem!(sys, subsystem_name, component.device)
+    end
+end
+
+"""
+Return a Generator of all components in the subsystem.
+
```

Comment:
Are we missing some code here?

Reply:
Management of subsystems occurs in the InfrastructureSystems (IS) layer. We implement storage of components and related objects in this sub-package so that other packages can benefit from the logic. The real function is here: https://github.com/daniel-thom/InfrastructureSystems.jl/blob/subsystems/src/subsystems.jl#L74

Reply:
Thank you!